uktrade/data-hub-frontend
# Upgrade to webpack v4

## Description of change

This PR updates webpack (from v3 to v4) and related dependencies to fix the below error when trying to run `npm run storybook`. 

To do this I started fresh with the webpack config and built it back up again piece by piece. Please let me know if you see anything that is missing. Some things will be different because behaviour/the APIs have changed. 

For example:

- `extract-text-webpack-plugin` has been deprecated in favour of `mini-css-extract-plugin`
- `UglifyJSPlugin` is no longer need as it is the default behaviour - https://webpack.js.org/migrate/4/#deprecatedremoved-plugins

I've also combined the different env configs into a single one as there are only a few things that change depending on the env and so it seemed to be easier to comprehend in one place. 

Our previous config also had a `webpack.config.docker.js`, but I couldn't see that it would be used ever, so I have removed it. But please let me know if this is in error. 

We've also updated some of the visual test baseline images. The differences here are minuscule. 



```
10% building 1/1 modules 0 active(node:11438) DeprecationWarning: Tapable.apply is deprecated. Call apply on the plugin directly instead
(node:11438) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'createSnapshot' of undefined
    at /Users/davidmatthews/code/DIT/datahub-front-end/node_modules/html-webpack-plugin/lib/webpack5/file-watcher-api.js:13:36
    at new Promise (<anonymous>)
    at Object.createSnapshot (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/html-webpack-plugin/lib/webpack5/file-watcher-api.js:12:10)
    at /Users/davidmatthews/code/DIT/datahub-front-end/node_modules/html-webpack-plugin/lib/cached-child-compiler.js:219:35
(node:11438) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
(node:11438) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/webpack/lib/SetVarMainTemplatePlugin.js:17
                compilation.templatesPlugin("render-with-entry", (source, chunk, hash) => {
                            ^
TypeError: compilation.templatesPlugin is not a function
    at SetVarMainTemplatePlugin.apply (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/webpack/lib/SetVarMainTemplatePlugin.js:17:15)
    at Compilation.apply (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/tapable/lib/Tapable.js:79:16)
    at Compilation.deprecated [as apply] (internal/util.js:89:15)
    at /Users/davidmatthews/code/DIT/datahub-front-end/node_modules/webpack/lib/LibraryTemplatePlugin.js:45:18
    at SyncHook.eval [as call] (eval at create (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:9:1)
    at SyncHook.lazyCompileHook (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/tapable/lib/Hook.js:154:20)
    at Compiler.newCompilation (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/@storybook/core/node_modules/webpack/lib/Compiler.js:630:30)
    at /Users/davidmatthews/code/DIT/datahub-front-end/node_modules/@storybook/core/node_modules/webpack/lib/Compiler.js:667:29
    at eval (eval at create (/Users/davidmatthews/code/DIT/datahub-front-end/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at /Users/davidmatthews/code/DIT/datahub-front-end/node_modules/@storybook/core/node_modules/webpack/lib/DllReferencePlugin.js:72:15
    at /Users/davidmatthews/code/DIT/datahub-front-end/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:85:15
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! data-hub-frontend@ storybook: `start-storybook`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the data-hub-frontend@ storybook script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/davidmatthews/.npm/_logs/2020-07-16T14_46_42_405Z-debug.log
```


## Test instructions

Everything should work as before and `npm run storybook` should also work. 

Please check that everything works in IE, in particular. 

Also, please check the JS bundle size to make sure this doesn't have any negative impact on performance. 

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"
